language-plutus-core/src/Language/PlutusCore/Normalize.hs

```diff
@@ -1,58 +1,130 @@
+-- | Environments, type normalization runs in.
+data NormalizeTypeEnv m tyname ann = NormalizeTypeEnv
+    { _normalizeTypeEnvTypeArgEnv :: TypeArgEnv tyname ann
+    , _normalizeTypeEnvCountStep  :: m ()
+      -- ^ How to count a type normalization step.
+    }
 
-type GasInit = Maybe Natural
+makeLenses ''NormalizeTypeEnv
 
-normalizeTypeStep :: NormalizeTypeM tyname a ()
-normalizeTypeStep = do
-    st <- get
-    case st of
-        Just 0  -> throwError OutOfGas
-        Just _  -> modify (fmap (subtract 1))
-        Nothing -> pure ()
+{- Note [NormalizeTypeT]
+Type normalization requires 'Quote' (because we need to be able to generate fresh names), but we
+do not put 'Quote' into 'NormalizeTypeT'. The reason for this is that it makes type signatures of
+various runners much nicer and also more generic. For example, we have
 
--- | Run a 'NormalizeTypeM' computation.
-runNormalizeTypeM :: (MonadQuote m, AsTypeError e a, MonadError e m) => GasInit -> NormalizeTypeM tyname a b -> m b
-runNormalizeTypeM mn a = throwingEither _TypeError =<< (liftQuote $ runExceptT $ runReaderT (evalStateT a mn) (TypeEnv mempty))
+    runNormalizeTypeAnyM :: MonadQuote m => NormalizeTypeT m tyname ann a -> m a
+
+If 'NormalizeTypeT' contained 'Quote', it would be
 
--- | Locally extend a 'TypeEnv' in a 'NormalizeTypeM' computation.
-withExtendedTypeEnv
-    :: HasUnique (tyname ()) TypeUnique
-    => tyname () -> NormalizedType tyname () -> NormalizeTypeM tyname ann a -> NormalizeTypeM tyname ann a
-withExtendedTypeEnv name ty =
-    local (TypeEnv . IntMap.insert (name ^. unique . coerced) ty . unTypeEnv)
+    runNormalizeTypeAnyM :: NormalizeTypeT m tyname ann a -> QuoteT m a
+
+which hardcodes 'QuoteT' to be the outermost transformer.
+
+Type normalization can run in any @m@ (as long as it's a 'MonadQuote') as witnessed by
+the following type signature:
+
+    normalizeTypeM
+        :: (HasUnique (tyname ann) TypeUnique, MonadQuote m)
+        => Type tyname ann -> NormalizeTypeT m tyname ann (NormalizedType tyname ann)
+
+so it's natural to have runners that do not break this genericity.
+-}
 
--- | Look up a @tyname@ in a 'TypeEnv'.
+-- See Note [NormalizedTypeT].
+-- | The monad transformer, type normalization runs in.
+newtype NormalizeTypeT m tyname ann a = NormalizeTypeT
+    { unNormalizeTypeT :: ReaderT (NormalizeTypeEnv m tyname ann) m a
+    } deriving newtype
+        ( Functor, Applicative, Alternative, Monad, MonadPlus
+        , MonadReader (NormalizeTypeEnv m tyname ann), MonadState s
+        , MonadQuote
+        )
+
+type CountSubstT m = StateT Gas (MaybeT m)
+
+-- | Run a 'NormalizeTypeM' computation.
+runNormalizeTypeM :: m () -> NormalizeTypeT m tyname ann a -> m a
+runNormalizeTypeM countStep (NormalizeTypeT a) =
+    runReaderT a $ NormalizeTypeEnv (TypeArgEnv mempty) countStep
+
+-- | Run a 'NormalizeTypeM' computation without dealing with gas.
+runNormalizeTypeAnyM
```

Comment:
Me too. Suggestions? `NoGas` instead of `Any`?

Reply:
`runNormalizeType` without the `M`, since this is in some sense the "non-monadic" version?

Reply:
We have `normalizeType` that receives a `m ()`. I.e. the default name requires you to say how you want to handle gas problems. If `normalizeType` simply normalizes a type, this can result in gotchas like "oh, I can't just normalize types, I need to think about gas".

I renamed the function to `runNormalizeTypeDownM`. I think it's better to be explicit in the name of a function that can be very expensive to compute.